lib/shelf_pubserver.dart

```diff
@@ -360,60 +352,59 @@ class ShelfPubServer {
   }
 
   Future<shelf.Response> _uploadSimple(
-      Uri uri, String contentType, Stream<List<int>> stream) {
+      Uri uri, String contentType, Stream<List<int>> stream) async {
     _logger.info('Perform simple upload.');
 
-    var mediaType = new MediaType.parse(contentType);
-
-    if (mediaType.type == 'multipart' && mediaType.subtype == 'form-data') {
-      var boundary = mediaType.parameters['boundary'];
-
-      // We have to listen to all multiparts: Just doing `parts.first` will
-      // result in the cancelation of the subscription which causes
-      // eventually a destruction of the socket, this is an odd side-effect.
-      // What we would like to have is something like this:
-      //     parts.expect(1).then((part) { upload(part); })
-      bool firstPartArrived = false;
-      var completer = new Completer<shelf.Response>();
-      StreamSubscription subscription;
-
-      var parts = stream.transform(new MimeMultipartTransformer(boundary));
-      subscription = parts.listen((MimeMultipart part) {
-        // If we get more than one part, we'll ignore the rest of the input.
-        if (firstPartArrived) {
-          subscription.cancel();
-          return;
-        }
-        firstPartArrived = true;
-
-        // TODO: Ensure that `part.headers['content-disposition']` is
-        // `form-data; name="file"; filename="package.tar.gz`
-        repository.upload(part).then((PackageVersion version) async {
-          if (cache != null) {
-            _logger
-                .info('Invalidating cache for package ${version.packageName}.');
-            await cache.invalidatePackageData(version.packageName);
-          }
-          _logger.info('Redirecting to found url.');
-          return new shelf.Response.found(_finishUploadSimpleUrl(uri));
-        }).catchError((String error, stack) {
-          _logger.warning('Error occured: $error\n$stack.');
-          // TODO: Do error checking and return error codes?
-          return new shelf.Response.found(
-              _finishUploadSimpleUrl(uri, error: error));
-        }).then(completer.complete);
-      });
+    var boundary = _getBoundary(contentType);
+
+    if (boundary == null) {
+      return _badRequest(
+          'Upload must contain a multipart/form-data content type.');
+    }
+
+    // We have to listen to all multiparts: Just doing `parts.first` will
+    // result in the cancellation of the subscription which causes
+    // eventually a destruction of the socket, this is an odd side-effect.
+    // What we would like to have is something like this:
+    //     parts.expect(1).then((part) { upload(part); })
+    MimeMultipart thePart;
+    StreamSubscription subscription;
+
+    var parts = stream.transform(new MimeMultipartTransformer(boundary));
+    subscription = parts.listen((MimeMultipart part) {
```

Comment:
You could then also use `await for() ` here, which would increase the readability.